functorch/_src/aot_autograd.py

```diff
@@ -286,7 +286,58 @@ def new_fn(args):
 
 
 def aot_dispatch_autograd(flat_fn, flat_args: List[Tensor], aot_config: AOTConfig):
-    joint_forward_backward = create_joint_forward_backward(flat_fn)
+    # Deduplicate inputs.  Suppose you have:
+    #
+    #   [a, b, a, c]
+    #
+    # We want:
+    #
+    #   remove_dupe_args([a, b, a, c]) == [a, b, c]
+    #   add_dupe_args([a, b, c]) == [a, b, a, c]
+    #
+    # This is done via (respectively):
+    #
+    #   seen_args = {2}  # what to drop
+    #   add_dupe_map = {  # how to get args from the deduped list
+    #       0: 0,
+    #       1: 1,
+    #       2: 0,
+    #       3: 2,
+    #   }
+    #
+    # Whether to use flat_args or deduped_flat_args?  flat_fn takes flat_args,
+    # and the autograd.Function must take deduped_flat_args; everything
+    # else is just getting the types right.
+
+    seen_args = {}
+    deduped_flat_args = []
+    drop_args = set()
+    add_dupe_map = {}
+    duped_arg_len = len(flat_args)
+    for i, t in enumerate(flat_args):
+        if t in seen_args:
+            drop_args.add(i)
+            add_dupe_map[i] = seen_args[t]
+            continue
+        seen_args[t] = len(deduped_flat_args)
+        add_dupe_map[i] = len(deduped_flat_args)
+        deduped_flat_args.append(t)
+
+    def remove_dupe_args(args):
+        r = []
+        for i, t in enumerate(args):
+            if i in drop_args:
```

Comment:
nit, since this is on the hot path, can we do something like 

```
for t, keep_arg in zip(args, keep_args):
    if keep_arg:
         r.append(t)
```